### PR TITLE
[MLIR] Handle tf.variant in tf-tensor-list-ops-decomposition pass

### PR DESCRIPTION
**Pull request by avarmapml**
-- Motivation : Currently `tf-tensor-list-ops-decomposition` pass requires full
                shape of the tensor list to be inferred.
                It bails out on tf.variant and tf.variant<tf.variant> where the
                max size and shape of the element is unknown and not immediately
                apparent.

-- This commit aims to handle tf.variant and tf.variant<tf.variant>.
   It does so by finding out the maximum number of elements being pushed in the
   empty tensor list and the size of each element.

-- It currently handles two cases:
   1. Variant declared by an empty tensor list op to which tensors are pushed.
   2. Variant declared by an empty tensor list op to which other variant declared
      by a reserve op is pushed.

Signed-off-by: Abhishek Varma <abhishek.varma@polymagelabs.com>

### CODE REVIEW

**Comment by avarmapml** [comment]
Hi @joker-eph @sanjoy @sherhut 
Can you please review this PR?

**Comment by sanjoy** [comment]
I'm not the correct reviewer for this.  I've pinged folks internally to help find an appropriate reviewer.

**Comment by joker-eph** [comment]
@smit-hinsu  said they'd look into it.

**Comment by avarmapml** [comment]
Thanks for responding @sanjoy and @joker-eph .
Awaiting @smit-hinsu 's review comments.

File: tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc [hunk 1]
```diff
@@ -574,9 +864,38 @@ LogicalResult HandleEmptyTensorListOp(
...
     }
   }
-  if (failed(cutil::CreateInitBufferValue(
-          element_shape, list.max_num_elements(), list, list.element_dtype(),
-          builder, &buffer))) {
+  Value list_size = list.max_num_elements();
+  auto list_size_op = list_size.getDefiningOp<TF::ConstOp>();
+  if (!list_size_op) return list.emitOpError("unknown max element count");
+  ElementsAttr listSizeAttr = list_size_op.value();
+  auto moduleOp = static_cast<Operation*>(list)->getParentOfType<ModuleOp>();
```
Comment (smit-hinsu) [inline tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc:871]: There is implicit cast so you should be able to remove explicit cast here and just rely on the cast to happen at the callsite.

File: tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc [hunk 2]
```diff
@@ -559,6 +560,295 @@ LogicalResult GetElementShapeFromResultType(
...
+    }
+  }
+  return false;
+}
+
+/// Returns the maximum number of tensor elements that'll be pushed in the
+/// EmptyTensorList. In case more than one push_back operation is applied on
+/// the EmptyTensorList within the same WhileOp body, we return None and thus
+/// the rewriter fails in those cases.
+static Optional<unsigned> findDynamicShape(TF::EmptyTensorListOp op,
```
Comment (smit-hinsu) [inline tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc:639]: What happens if there are ops like TensorListPushBackBatch, TensorListConcatV2 and others? There might be some new ops in the future or some custom tensor list ops by the users.

**Review by smit-hinsu — COMMENTED** [review]
Could you also expand the commit description to include motivation?

File: tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc (hunk 2 above)
Comment (avarmapml) [reply tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc:639]: That's a very good point @smit-hinsu .
Currently the code only deals with `TensorListPushBack` ops with an added constraint that only 1 push back per iteration is allowed : `checkExactlyOnePushbackUse`.

The constraint can be relaxed gradually in subsequent PRs and I've added a TODO for handling `TensorListPushBackBatch`, `TensorListConcatV2`, etc. One way to deal with them in the subsequent PRs is to perhaps add an OpTrait in such ops that changed the shape of the list and go about with it.

**Comment by avarmapml** [comment]
> Could you also expand the commit description to include motivation?

Hi @smit-hinsu I've made the appropriate changes and included the motivation in commit summary.

File: tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc (hunk 2 above)
Comment (smit-hinsu) [reply tensorflow/compiler/mlir/tensorflow/transforms/tensor_list_ops_decomposition.cc:643]: That will make this transformation incorrect. Trait can handle existing operations but someone could be using custom operations that we don't know about.

**Comment by smit-hinsu** [comment]
Thanks for adding motivation. That is useful. Look at compiler/mlir/tensorflow/transforms/shape_inference.cc pass that does some shape inference which you might be able to utilize. ShapeInference::InferShapeForTensorListInitOps and around code.

**Comment by gbaned** [comment]
@avarmapml Can you please check @smit-hinsu's comments and keep us posted ? Thanks!

**Comment by gbaned** [comment]
@avarmapml  Any update on this PR? Please. Thanks!

**Comment by avarmapml** [comment]
Apologies for the delay.
I tried exploring ShapeInference::InferShapeForTensorListInitOps but what I understand from it is :- 
1. It tries inferring the basic elemental shape of TensorList ops. eg: `tf.variant<?x1xf32>` -> `tf.variant<16x1xf32>`. We can perhaps sequence the passes like : `-tf-shape-inference -tf-tensor-list-ops-decomposition` in order to further deal with these ops.
2. Thus the state of the IR where `tf-shape-inference` lands us is where `tf-tensor-list-ops-decomposition` begins.

Can you tell me what else can be done here? @smit-hinsu

**Comment by gbaned** [comment]
@smit-hinsu  Can you please assist on above comments from @avarmapml. Thanks!

**Comment by vinayaka-polymage** [comment]
@smit-hinsu Can you please check this when you get a chance ?

**Comment by gbaned** [comment]
@smit-hinsu  Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@avarmapml Any update on this PR? Please. Thanks!

**Pull request closed without merging**